kale/embed/_mpca_torch.py

```diff
@@ -0,0 +1,236 @@
+            cum = eig_values[idx_sorted]
+            var_tot = sum(cum)
+
+            for j in range(1, cum.shape[0] + 1):
+                if sum(cum[:j]) / var_tot > self.var_ratio:
+                    shape_out += (j,)
+                    break
+            # cums[i] = cum
+            self.factors.append(eig_vectors[:, idx_sorted][:, : shape_out[i - 1]].T)
+
+        all_comps = int(torch.prod(torch.Tensor(shape_out)))
+        self.idx_order = range(all_comps)
+
+        for i_iter in range(self.max_iter):
+            phi = dict()
+            for i in range(1, n_dim):  # ith mode
+                if i not in phi:
+                    phi[i] = 0
+                for j in range(n_spl):
+                    x_j = x[j, :]  # jth tensor/sample
+                    xj_ = multi_mode_dot(
+                        x_j,
+                        [self.factors[m] for m in range(n_dim - 1) if m != i - 1],
+                        modes=[m for m in range(n_dim - 1) if m != i - 1],
+                    )
+                    xj_unfold = tl.unfold(xj_, i - 1)
+                    phi[i] = torch.matmul(xj_unfold, xj_unfold.T) + phi[i]
+
+                eig_values, eig_vectors = torch.eig(phi[i], eigenvectors=True)
+                eig_values = eig_values[:, 0]
+                idx_sorted = (-1 * eig_values).argsort()
+                # lambdas[i] = eig_values[idx_sorted]
+                self.factors[i - 1] = (eig_vectors[:, idx_sorted][:, : shape_out[i - 1]]).T
+
+        tensor_pc_vec = self.forward(x, reduce_mean=False, return_vector=True, sort_by_var=False)
+        # return_vector the tensor principal components
+        # diagonal of the covariance of vectorised features
+        tensor_pc_cov_diag = torch.diag(torch.matmul(tensor_pc_vec.T, tensor_pc_vec))
+        idx_order = (-1 * tensor_pc_cov_diag).argsort()
+
+        self.idx_order = idx_order
+        self.shape_out = shape_out
+
+        return self
+
+    def training_step(self, train_batch):
+        """Training"""
+        x, y = train_batch
+        self._fit(x)
+        return self
+
+    def forward(self, x, reduce_mean=True, return_vector=False,
+                sort_by_var=True, n_components=None):
+        """Perform dimension reduction on x
+
+        Args:
+            x (tensor):
+            reduce_mean (bool): Whether reduce the mean estimated from training data. Defaults to True.
+            return_vector (bool): Whether return the tensor principal components in vector.
+                Defaults to False.
+            sort_by_var (bool): Whether sort features by variance in descending order. Defaults to True,
+                applies only when return_vector=True.
+            n_components (int): Number of components to keep. Applies only when return_vector=True.
+                Defaults to None.
+
+        Returns:
+            tensor
+                tensor principal components, shape (n_samples, P_1, P_2, ..., P_N) if return_vector==False.
```

Comment:
I think you confuse the PCs with PCA features/projections. Read the first sentence of [PCA Wiki](https://en.wikipedia.org/wiki/Principal_component_analysis) to learn what the PCs, they are the bases.

Reply:
I suggest you to use bases and projections if "features" is confusing.

Reply:
How about principal component bases (pc_bases) and tensor projections (tensor_proj)?

Reply:
Will tensor_proj be confused with the projection matrices (which you called factors in one and proj_mats in another).
I try to learn from wiki. Maybe for the low-dimensional representations, we call it `projected_tensor` (see par 2 of wiki pca).
For the bases, you already have a name called `proj_mats`, which may be better as `projection_mats`. What will your `pc_bases` for? We can discuss in our meeting.